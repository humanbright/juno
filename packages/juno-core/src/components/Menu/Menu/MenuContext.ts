import { createContext } from 'react';

export type RcMenuOnCloseReasonsType =
  | 'backdropClick'
  | 'escapeKeyDown'
  | 'tabKeyDown'
  | 'itemClick';

export type RcMenuContextType = {
  menuId: string;
  autoClose: boolean;
  focusedMenuItemId: string | null;
  setFocusedMenuItemId: React.Dispatch<React.SetStateAction<string | null>>;
  closeMenu: (event: {}, reason: RcMenuOnCloseReasonsType) => void;
};

export const RcMenuContext = createContext<RcMenuContextType>({
  menuId: '',
  autoClose: false,
  focusedMenuItemId: null,
  setFocusedMenuItemId: () => null,
  closeMenu: () => null,
});
