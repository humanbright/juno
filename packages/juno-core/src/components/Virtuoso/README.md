# react-virtuoso

This project fork from [react-virtuoso](https://github.com/petyosi/react-virtuoso.git)

v2.8.6
SHA: 02f188051767d52e47a527bbdce39f61a140a51e

always keep sync with original source, we extends that for support custom `window instance`.
