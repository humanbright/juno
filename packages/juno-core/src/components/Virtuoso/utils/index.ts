export * from './isOutOfRange';
export * from './useDynamicHeight';
export * from './useHighlightScroll';
