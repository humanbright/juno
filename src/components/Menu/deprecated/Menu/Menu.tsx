import MuiMenu from '@material-ui/core/Menu';
import React, {
  ComponentProps,
  forwardRef,
  memo,
  useMemo,
  useState,
} from 'react';

import {
  RcBaseProps,
  styled,
  useEventCallback,
  useRcPortalWindowContext,
  withDeprecatedCheck,
} from '../../../../foundation';
import { StyledMenu } from './styles';

type RcMenuProps = {
  autoClose?: boolean;
  /** @deprecated use autoClose to replace that */
  enableCloseMenuWhenItemClicked?: boolean;
} & RcBaseProps<ComponentProps<typeof MuiMenu>>;

type CloseReasonType = 'backdropClick' | 'escapeKeyDown' | 'tabKeyDown';

type MenuContextType = {
  autoClose: boolean;
  open: boolean;
  closeMenu: (event: {}, reason: CloseReasonType) => void;
  focusedMenuItemId: string;
  setFocusedMenuItemId: (id: string) => void;
};

const MenuContext = React.createContext<MenuContextType>({
  autoClose: false,
  open: false,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  closeMenu: (event: {}, reason: CloseReasonType) => {},
  focusedMenuItemId: '',
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  setFocusedMenuItemId: (id: string) => {},
});

// TODO: should use ref to refactor this menu logic with focus item set
const MemoMenu = memo(StyledMenu);

const _RcMenu = forwardRef<any, RcMenuProps>(
  (
    {
      onClose,
      open,
      enableCloseMenuWhenItemClicked,
      autoClose = enableCloseMenuWhenItemClicked,
      children,
      ...rest
    },
    ref,
  ) => {
    const [focusedMenuItemId, setFocusedMenuItemId] = useState('');
    const { externalWindow } = useRcPortalWindowContext();

    const handleClose = useEventCallback(
      (event: {}, reason: CloseReasonType) => {
        onClose?.(event, reason as Exclude<CloseReasonType, 'tabKeyDown'>);
      },
    );

    const ctxValue: MenuContextType = useMemo(
      () => ({
        autoClose: !!autoClose,
        open,
        closeMenu: handleClose,
        focusedMenuItemId,
        setFocusedMenuItemId: (id: string) => setFocusedMenuItemId(id),
      }),
      [autoClose, focusedMenuItemId, handleClose, open],
    );

    return (
      <MenuContext.Provider value={ctxValue}>
        <MemoMenu
          container={externalWindow?.document.body}
          {...rest}
          open={open}
          ref={ref}
          onClose={handleClose}
          disableEnforceFocus
        >
          {children}
        </MemoMenu>
      </MenuContext.Provider>
    );
  },
);

/** @release */
const RcMenu = styled(
  withDeprecatedCheck(
    _RcMenu,
    [
      {
        prop: 'enableCloseMenuWhenItemClicked',
        time: '2021-2',
        comment: `use autoClose to replace that `,
      },
    ],
    'RcMenu',
  ),
)``;

RcMenu.displayName = 'RcMenu';

RcMenu.defaultProps = {
  getContentAnchorEl: null,
  anchorOrigin: { vertical: 'bottom', horizontal: 'left' },
};

export { RcMenu, RcMenuProps, MenuContext as RcMenuContext };
