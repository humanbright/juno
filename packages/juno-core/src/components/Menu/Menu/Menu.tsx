import React, {
  ComponentProps,
  forwardRef,
  memo,
  useMemo,
  useState,
} from 'react';

import MuiMenu from '@material-ui/core/Menu';

import {
  combineClasses,
  RcBaseProps,
  styled,
  useEventCallback,
  useId,
  useRcPortalWindowContext,
  useThemeProps,
} from '../../../foundation';
import type {
  RcMenuContextType,
  RcMenuOnCloseReasonsType,
} from './MenuContext';
import { RcMenuContext } from './MenuContext';
import { MenuStyle } from './styles';
import { RcMenuClasses } from './utils';

type RcMenuProps = {
  /** auto close menu when menu item clicked */
  autoClose?: boolean;
  /** emit when menu closed */
  onClose?: (event: {}, reason: RcMenuOnCloseReasonsType) => void;
} & RcBaseProps<
  ComponentProps<typeof MuiMenu>,
  'disableEnforceFocus' | 'onClose'
>;

const MemoMuiMenu = memo(MuiMenu);

const _RcMenu = forwardRef<any, RcMenuProps>((inProps: RcMenuProps, ref) => {
  const props = useThemeProps({ props: inProps, name: 'RcMenu' });
  const { classes: classesProp, children, autoClose, onClose, ...rest } = props;
  const [focusedMenuItemId, setFocusedMenuItemId] = useState(null);
  const id = useId('menu-', true);

  const { externalWindow } = useRcPortalWindowContext();

  const classes = useMemo(
    () => combineClasses(RcMenuClasses, classesProp),
    [classesProp],
  );

  const handleClose = useEventCallback(
    (event: {}, reason: RcMenuOnCloseReasonsType) => {
      onClose?.(event, reason);
    },
  );

  const ctxValue: RcMenuContextType = useMemo(
    () => ({
      menuId: id,
      autoClose: !!autoClose,
      focusedMenuItemId,
      setFocusedMenuItemId,
      closeMenu: handleClose,
    }),
    [autoClose, focusedMenuItemId, handleClose, id],
  );

  return (
    <RcMenuContext.Provider value={ctxValue}>
      <MemoMuiMenu
        container={externalWindow?.document.body}
        {...rest}
        ref={ref}
        classes={classes}
        onClose={handleClose}
        disableEnforceFocus
      >
        {children}
      </MemoMuiMenu>
    </RcMenuContext.Provider>
  );
});

const RcMenu = styled(_RcMenu)`
  ${MenuStyle};
`;

RcMenu.defaultProps = {
  variant: 'selectedMenu',
  getContentAnchorEl: null,
  anchorReference: 'anchorEl',
  anchorOrigin: { vertical: 'bottom', horizontal: 'left' },
};

RcMenu.displayName = 'RcMenu';

export { RcMenu, RcMenuContext };

export type { RcMenuContextType, RcMenuProps };
