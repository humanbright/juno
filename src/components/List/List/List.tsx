import MuiList from '@material-ui/core/List';
import React, { ComponentProps, forwardRef, useMemo } from 'react';

import {
  combineClasses,
  RcBaseProps,
  styled,
  useThemeProps,
} from '../../../foundation';
import { ListStyle } from './styles';
import { RcListClasses } from './utils';

type RcListProps = {} & RcBaseProps<ComponentProps<typeof MuiList>>;

const _RcList = forwardRef<any, RcListProps>((inProps, ref) => {
  const props = useThemeProps({ props: inProps, name: 'RcList' });
  const { classes: classesProp, children, ...rest } = props;
  const classes = useMemo(
    () => combineClasses(RcListClasses, classesProp),
    [classesProp],
  );

  return (
    <MuiList {...rest} ref={ref} classes={classes}>
      {children}
    </MuiList>
  );
});

const RcList = styled(_RcList)`
  ${ListStyle}
`;

RcList.defaultProps = {
  disablePadding: true,
};

RcList.displayName = 'RcList';

export { RcList };
export type { RcListProps };
