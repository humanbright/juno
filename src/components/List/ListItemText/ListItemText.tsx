import MuiListItemText from '@material-ui/core/ListItemText';
import React, { ComponentProps, forwardRef, useMemo } from 'react';

import {
  combineClasses,
  omit,
  RcBaseProps,
  RcPaletteProp,
  styled,
  withDeprecatedCheck,
  useThemeProps,
} from '../../../foundation';
import { RcTypographyProps } from '../../Typography';
import { ListItemTextStyle } from './styles';
import { RcListItemTextClasses } from './utils';

type TypographyProps = RcBaseProps<RcTypographyProps, 'weight'>;

type RcListItemTextProps = {
  /** is text Ellipsis, default is true. */
  isEllipsis?: boolean;
  /** These props will be forwarded to the primary typography component (as long as disableTypography is not true). */
  primaryTypographyProps?: TypographyProps;
  /** These props will be forwarded to the secondary typography component (as long as disableTypography is not true). */
  secondaryTypographyProps?: TypographyProps;
  /** @deprecated define color for primary, recommend use primaryTypographyProps color prop */
  primaryColor?: RcPaletteProp;
  /** @deprecated is alignCenter, recommend using classes to define it. */
  alignCenter?: boolean;
  /** @deprecated is lineThrough, recommend using classes to define it. */
  lineThrough?: boolean;
} & RcBaseProps<
  ComponentProps<typeof MuiListItemText>,
  'primaryTypographyProps' | 'secondaryTypographyProps'
>;

const _RcListItemText = forwardRef<any, RcListItemTextProps>((inProps, ref) => {
  const props = useThemeProps({ props: inProps, name: 'RcListItemText' });
  const {
    classes: classesProp,
    children,
    isEllipsis,
    primaryTypographyProps: primaryTypographyPropsProp,
    secondaryTypographyProps: secondaryTypographyPropsProp,
    primaryColor,
    alignCenter,
    lineThrough,
    ...rest
  } = props;

  const classes = useMemo(
    () => combineClasses(RcListItemTextClasses, classesProp),
    [classesProp],
  );

  const primaryTypographyProps = useMemo(
    () => omit(primaryTypographyPropsProp, ['color', 'variant']),
    [primaryTypographyPropsProp],
  );

  const secondaryTypographyProps = useMemo(
    () => omit(secondaryTypographyPropsProp, ['color', 'variant']),
    [secondaryTypographyPropsProp],
  );

  return (
    <MuiListItemText
      {...rest}
      ref={ref}
      classes={classes}
      primaryTypographyProps={primaryTypographyProps}
      secondaryTypographyProps={secondaryTypographyProps}
    >
      {children}
    </MuiListItemText>
  );
});

const RcListItemText = styled(
  withDeprecatedCheck(
    _RcListItemText,
    [
      {
        prop: 'primaryColor',
        time: '2021-4',
        comment: 'recommend use primaryTypographyProps color prop',
      },
      {
        prop: 'alignCenter',
        time: '2021-4',
        comment: `recommend using classes to define it`,
      },
      {
        prop: 'lineThrough',
        time: '2021-4',
        comment: `recommend using classes to define it`,
      },
    ],
    'RcListItemText',
  ),
)`
  ${ListItemTextStyle}
`;

RcListItemText.defaultProps = {
  isEllipsis: true,
  primaryTypographyProps: {
    variant: 'body1',
    color: 'neutral.f06',
  },
  secondaryTypographyProps: {
    variant: 'caption1',
    color: 'neutral.f04',
  },
};

RcListItemText.displayName = 'RcListItemText';

export { RcListItemText };
export type { RcListItemTextProps };
