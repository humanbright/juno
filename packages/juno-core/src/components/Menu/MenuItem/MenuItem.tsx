import React, {
  ComponentProps,
  ElementType,
  forwardRef,
  ReactElement,
  ReactNode,
  Ref,
  useContext,
  useImperativeHandle,
  useMemo,
} from 'react';

import clsx from 'clsx';

import MuiMenuItem from '@material-ui/core/MenuItem';

import {
  Classes,
  combineClasses,
  isRcElement,
  logInDev,
  omit,
  RcBaseProps,
  RcBaseSize,
  RcClassesProps,
  styled,
  useEventCallback,
  useId,
  useThemeProps,
} from '../../../foundation';
import { Check as checkSvg } from '@ringcentral/juno-icon';
import { RcIcon, RcIconProps } from '../../Icon';
import { RcListItemSecondaryAction } from '../../List';
import { RcListItemAvatar } from '../../List/ListItemAvatar';
import { RcListItemIcon } from '../../List/ListItemIcon';
import { withTooltip, WithTooltipProps } from '../../Tooltip';
import { RcMenuContext } from '../Menu/MenuContext';
import { RcMenuListContext } from '../MenuList/MenuListContext';
import { RcSubMenuContext } from '../SubMenu/SubMenuContext';
import { MenuItemStyle, StyledCheckIcon } from './styles';
import { RcMenuItemClasses } from './utils';

type RcMenuItemSize = RcBaseSize<'large' | 'medium'>;

type RcMenuItemType = 'checked' | 'selected';

type RcMenuItemClassesType = RcClassesProps<
  Classes<ComponentProps<typeof MuiMenuItem>> | 'checked' | 'unchecked'
>;

type RcMenuItemInnerProps = {
  /** id ref, can get menu item id */
  idRef?: Ref<string>;
  /** for subMenu component, use prevent click */
  isSubMenuItem?: boolean;
};

type RcMenuItemProps = {
  /** MenuItem size */
  size?: RcMenuItemSize;
  /** The component used for the root node. Either a string to use a HTML element or a component. */
  component?: ElementType;
  /** checked for that should render checked icon */
  type?: RcMenuItemType;
  /** if type to be `checked`, the checked prop for that should render checked icon */
  checked?: boolean;
  /** MenuItem with icon, can use `RcListItemIcon` */
  icon?: ReactNode;
  /** pass to the icon, can use RcIcon with symbol prop */
  symbol?: RcIconProps['symbol'];
  /** MenuItem with avatar, can use ListItemAvatar */
  avatar?: ReactElement;
  /** MenuItem with subAction, can use ListItemSecondaryAction */
  secondaryAction?: ReactNode;
} & RcMenuItemClassesType &
  WithTooltipProps &
  RcBaseProps<ComponentProps<typeof MuiMenuItem>, 'classes' | 'title'>;

const _RcMenuItem = forwardRef<any, RcMenuItemProps & RcMenuItemInnerProps>(
  (inProps, ref) => {
    const props = useThemeProps({ props: inProps, name: 'RcMenuItem' });
    const {
      classes: classesProp,
      children,
      className,
      onMouseEnter,
      onClick,
      onFocus,
      size,
      type,
      checked,
      icon,
      symbol,
      avatar,
      secondaryAction,
      idRef,
      isSubMenuItem,
      title,
      ...rest
    } = props;

    const isCheckedType = type === 'checked';
    const menuItemId = useId('menu-item', true);

    const menuContext = useContext(RcMenuContext);
    const menuListContext = useContext(RcMenuListContext);
    const subMenuContext = useContext(RcSubMenuContext);

    const setFocusedMenuItemId = useEventCallback(() => {
      if (!menuItemId) return;

      if (menuListContext.menuListId) {
        menuListContext.setFocusedMenuItemId(menuItemId);
      } else if (menuContext.menuId) {
        menuContext.setFocusedMenuItemId(menuItemId);
      }
    });

    const classes = useMemo(
      () => combineClasses(RcMenuItemClasses, classesProp),
      [classesProp],
    );

    const toClasses = useMemo(
      () => omit(classes, ['checked', 'unchecked']),
      [classes],
    );

    const itemAvatar = useMemo(() => {
      if (React.isValidElement(avatar)) {
        return isRcElement(avatar, ['RcListItemAvatar']) ? (
          avatar
        ) : (
          <RcListItemAvatar>{avatar}</RcListItemAvatar>
        );
      }
      return null;
    }, [avatar]);

    const itemIcon = useMemo(() => {
      if (symbol) {
        return (
          <RcListItemIcon>
            <RcIcon size="small" symbol={symbol} />
          </RcListItemIcon>
        );
      }

      if (React.isValidElement(icon)) {
        return isRcElement(icon, ['RcListItemIcon']) ? (
          icon
        ) : (
          <RcListItemIcon>{icon}</RcListItemIcon>
        );
      }

      if (typeof icon === 'string') {
        logInDev({
          component: 'RcMenuItem',
          message: 'please use symbol to replace icon string',
        });
        return (
          <RcListItemIcon>
            <RcIcon size="small">{icon}</RcIcon>
          </RcListItemIcon>
        );
      }

      return null;
    }, [icon, symbol]);

    const itemSubAction = useMemo(() => {
      if (isCheckedType && checked) {
        return (
          <RcListItemSecondaryAction>
            <StyledCheckIcon
              size="large"
              symbol={checkSvg}
              color={['interactive', 'f01']}
            />
          </RcListItemSecondaryAction>
        );
      }

      return secondaryAction ? (
        isRcElement(secondaryAction, ['RcListItemSecondaryAction']) ? (
          secondaryAction
        ) : (
          <RcListItemSecondaryAction>
            {secondaryAction}
          </RcListItemSecondaryAction>
        )
      ) : null;
    }, [checked, isCheckedType, secondaryAction]);

    const handleMouseEnter = useEventCallback(
      (e: React.MouseEvent<HTMLLIElement>) => {
        onMouseEnter?.(e);
        setFocusedMenuItemId();
      },
    );

    const handleClick = useEventCallback(
      (e: React.MouseEvent<HTMLLIElement>) => {
        onClick?.(e);

        if (menuContext.autoClose && !isSubMenuItem) {
          menuContext.closeMenu(e, 'itemClick');
        }
        if (subMenuContext.autoClose && !isSubMenuItem) {
          subMenuContext?.closeSubMenu?.(e, 'itemClick');
        }
      },
    );

    const handleFocus = useEventCallback(
      (e: React.FocusEvent<HTMLLIElement>) => {
        setFocusedMenuItemId();
        onFocus?.(e);
      },
    );

    useImperativeHandle(idRef, () => menuItemId, [menuItemId]);

    return (
      <MuiMenuItem
        aria-checked={isCheckedType && checked ? true : undefined}
        {...rest}
        ref={ref as any}
        title={typeof title === 'string' ? title : undefined}
        classes={toClasses}
        className={clsx(className, {
          [classes!.checked]: isCheckedType && checked,
          [classes!.unchecked]: isCheckedType && !checked,
        })}
        onMouseEnter={handleMouseEnter}
        onClick={handleClick}
        onFocus={handleFocus}
        data-menuitem-id={menuItemId}
      >
        {itemAvatar}
        {itemIcon}
        {children}
        {itemSubAction}
      </MuiMenuItem>
    );
  },
);

const RcMenuItem = styled(withTooltip(_RcMenuItem))`
  ${MenuItemStyle};
`;

RcMenuItem.defaultProps = {
  size: 'medium',
};

RcMenuItem.displayName = 'RcMenuItem';

export { RcMenuItem };
export type {
  RcMenuItemClassesType,
  RcMenuItemProps,
  RcMenuItemSize,
  RcMenuItemType,
};
