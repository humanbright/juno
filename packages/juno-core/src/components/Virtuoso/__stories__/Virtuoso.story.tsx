import React, { ComponentProps } from 'react';

import range from 'lodash/range';

import {
  notControlInDocTable,
  notShowInDocTable,
  sortInDocTable,
} from '@ringcentral/juno-storybook';
import { Meta, Story } from '@storybook/react';

import { RcListItemText } from '../../List/ListItemText';
import { RcMenuItem } from '../../Menu/MenuItem';
import { Virtuoso } from '../Virtuoso';

export default {
  title: 'ðŸš€ Cleanup Components/Virtuoso',
  component: Virtuoso,
  argTypes: {
    ...sortInDocTable<keyof VirtuosoProps>([]),
    ...notControlInDocTable<keyof VirtuosoProps>([]),
    ...notShowInDocTable<keyof VirtuosoProps>([]),
  },
} as Meta;

type VirtuosoProps = ComponentProps<typeof Virtuoso>;

const menus = range(0, 1000);

export const VirtuosoExample: Story<Partial<VirtuosoProps>> = ({
  children,
  ...args
}) => {
  return (
    <>
      <Virtuoso
        style={{ height: '200px' }}
        totalCount={1000}
        data={menus}
        itemContent={(index) => {
          const label = `${index}-MenuItem`;
          return (
            <RcMenuItem key={index} data-search-text={label}>
              <RcListItemText primary={label} />
            </RcMenuItem>
          );
        }}
        {...args}
      />
    </>
  );
};

VirtuosoExample.args = {};

VirtuosoExample.argTypes = {
  ...notControlInDocTable<keyof VirtuosoProps>([]),
};

VirtuosoExample.parameters = {
  tags: [
    {
      name: 'Spec',
      href: 'https://share.goabstract.com/22b47ad0-05ea-4bcc-8bf0-37aafc4b38ae?mode=design&sha=9e531c6b30974c6e11ee1ca2030d489988c01e23',
    },
    {
      name: 'Mui',
      href: 'https://material-ui.com/components/menus/',
      value: 'Menu',
    },
    {
      name: 'Source',
      value: 'virtuoso',
      href: 'https://virtuoso.dev/',
      color: 'avatar.lake',
    },
    {
      name: 'Accessibility',
      value: '100%',
    },
  ],
};
