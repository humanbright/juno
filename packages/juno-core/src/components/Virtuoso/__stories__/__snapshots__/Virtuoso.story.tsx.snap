// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ðŸš€ Cleanup Components/Virtuoso Virtuoso Example 1`] = `
<div
  data-test-id="virtuoso-scroller"
  data-virtuoso-scroller={true}
  style={
    Object {
      "WebkitOverflowScrolling": "touch",
      "height": "200px",
      "outline": "none",
      "overflowY": "auto",
      "position": "relative",
    }
  }
  tabIndex={0}
>
  <div
    data-viewport-type="element"
    style={
      Object {
        "height": "100%",
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <div
      data-test-id="virtuoso-item-list"
      style={
        Object {
          "boxSizing": "border-box",
          "marginTop": 0,
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    />
  </div>
</div>
`;
